doc/website/blog/2022-03-14-ent-hooks-examples.md

```diff
@@ -0,0 +1,434 @@
+
+### So first, let us review in a few words what are hooks?
+
+[Hooks](https://entgo.io/docs/hooks) are a feature of Ent that allows adding custom logic before and after operations that change the data entities.
+
+A mutation is an operation that changes something in the database.
+There are 5 types of mutations:
+1. Create.
+2. UpdateOne.
+3. Update.
+4. DeleteOne.
+5. Delete.
+
+Hooks are functions that get an [ent.Mutator](https://pkg.go.dev/entgo.io/ent#Mutator) and return a mutator back.
+They function similar to the popular [HTTP middleware pattern](https://github.com/go-chi/chi#middleware-handlers).
+
+
+```go
+package example
+
+import (
+	"context"
+
+	"entgo.io/ent"
+)
+
+func exampleHook() ent.Hook {
+	//use this to init your hook
+	return func(next ent.Mutator) ent.Mutator {
+		return ent.MutateFunc(func(ctx context.Context, m ent.Mutation) (ent.Value, error) {
+			// Do something before mutation.
+			v, err := next.Mutate(ctx, m)
+			if err != nil {
+				// Do something if error after mutation.
+			}
+			// Do something after mutation.
+			return v, err
+		})
+	}
+}
+```
+
+In Ent, there are two types of mutation hooks - schema hooks and runtime hooks. Schema hooks are mainly used for defining
+custom mutation logic on a specific entity type, for example, syncing entity creation to another system. Runtime hooks, on the other hand, are used
+to define more global logic for adding things like logging, metrics, tracing, etc.
+
+In this post, I will focus mainly on schemas hooks.
+
+### So why use them?
+
+In ent, we try and push the entity's logic down to the schema level, this allows us to define everything about an entity close to its creation. Entity properties could be:
+1. Valid values.
+2. Types.
+3. Allowed mutations.
+4. Privacy.
+5. Required Side effect (heavy compute for example).
+
+---
+Many times, it’s more comfortable to define that in the RPC level, a controller provides a context and all the mutations
+and validation are done before changing the value.
+Yet, as the system gets bigger and more developers are working on it making sure that these validations and calculations are done
+whenever the entity is changing becomes harder and requires more pre-known knowledge about each entity to work with and
+can cause unintentional bugs or inconsistencies in the data that are only discovered in runtime.
+The way ent, based on the experience of bigger companies like Facebook, decided to handle it is to keep this information
+as close as possible to the entity, therefore hooks are just that, a piece of code that runs before and after an entity 
+is changing and defines that additional behavior.
+
+## Let us do some examples
+
+For this examples we use a simple schema of users and dogs.
+Each user can have multiple pets and each pet has an owner.
+We have also added a cache entity that will be used later.
+as can be seen here:
+```go
```

Comment:
add a  title to hint about the directory structure (especially for new users it might be confusing otherwise)